File: 2024/Golang/day2/part2.go
package day2

import (
	"Golang/utils"
	"bufio"
	"fmt"
	"os"
	"strings"
)

func Part2() {
	count := 0
	file, err := os.Open("../Input/day2.1.txt")
	utils.Check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		numbers := strings.Split(line, " ")
		report, err := utils.StrSliceToIntSlice(numbers)
		if err != nil {
			fmt.Println("Error parsing numbers:",err)
			continue
		}

		if utils.IsSafe(report) {
			count++
		} else {
			for i := 0; i < len(report); i++ {
				unsafeReports := append([]int64{}, report[:i]...)
				unsafeReports = append(unsafeReports, report[i + 1:]...)
				if utils.IsSafe(unsafeReports) {
					count++
					break
				}
			}
		}
	}
	fmt.Println(count)
}

File: 2024/Rust/day2/part2.rs
use crate::utils;

pub fn run_part2() {
    let mut safe_count = 0;
    for line in utils::read_lines("../Input/day2.1.txt") {
        let report: Vec<i64> = line.split(' ').map(|x| x.parse::<i64>().unwrap()).collect();
        if utils::is_safe(report.clone()) {
            safe_count += 1;
        } else {
            for i in 0..report.len() {
                let mut unsafe_reports = report.clone();
                unsafe_reports.remove(i);
                if utils::is_safe(unsafe_reports) {
                    safe_count += 1;
                    break;
                }
            }
        }
    }

    println!("{}", safe_count);
}